oidc-exchange.py

```diff
@@ -0,0 +1,65 @@
+import os
+import sys
+
+import requests
+
+
+def die(msg: str):
+    print(f"OIDC exchange failure: {msg}", file=sys.stderr)
+    sys.exit(1)
+
+
+oidc_audience = os.getenv("INPUT_OIDC_AUDIENCE")
+
+oidc_token_url = os.getenv("ACTIONS_ID_TOKEN_REQUEST_URL")
+if oidc_token_url is None:
+    die("no ACTIONS_ID_TOKEN_REQUEST_URL in environment")
+
+oidc_bearer_token = os.getenv("ACTIONS_ID_TOKEN_REQUEST_TOKEN")
+if oidc_bearer_token is None:
+    die("no ACTIONS_ID_TOKEN_REQUEST_TOKEN in environment")
+
+oidc_resp = requests.get(
+    oidc_token_url,
+    params={"audience": oidc_audience},
+    headers={"Authorization": f"bearer {oidc_bearer_token}"},
+)
+
+if not oidc_resp.ok:
+    die(f"OIDC credential request error: {oidc_resp.status_code}")
+
+oidc_payload = oidc_resp.json()
+oidc_token = oidc_payload.get("value")
+if oidc_token is None:
+    # We emit the OIDC payload here, knowing that GitHub Actions will mask
+    # the token contents (if present).
+    die(f"malformed OIDC credential response: {oidc_payload}")
+
+# Now we can do the actual token exchange.
+repository_url = os.getenv("INPUTS_REPOSITORY_URL")
+if repository_url is None:
+    # Easy case: no explicit repository URL, which means we're using PyPI and we can just
+    # hardcode the exchange endpoint.
+    token_exchange_url = "https://pypi.org/_/oidc/github/mint-token"
+else:
+    # Annoying case: the repository URL is specified explicitly, which means we need to extract
+    # its domain name and build up the endpoint from there.
+    # TODO: Actually do this.
+    die(f"unimplemented: token exchange for non-PyPI instances")
+
+mint_token_resp = requests.post(
+    token_exchange_url,
+    json={"token": oidc_token},
+)
+
+if not mint_token_resp.ok:
+    # TODO: Better error message here.
+    die(f"mint-token request error: {mint_token_resp.status_code}")
```

Comment:
I think we'll need to handle a couple possible sub-conditions here, with more appropriate error message:

- what PyPI gives back if OIDC is disabled
- what PyPI gives back if the user isn't in the beta
- what PyPI gives back if there is no publisher configured for this OIDC token
- catchall for any other possible failure response from a non-PyPI index

Reply:
Specifically for this one:

> what PyPI gives back if there is no publisher configured for this OIDC token

We'll want to make this seem less like an error and instead push the user to configure an OIDC publisher. This will probably be commonly encountered by people unaware that OIDC support exists, and essentially will replace the error condition that currently happens when the user forgot to provide a password/API token.

Reply:
Sounds good! Thanks for enumerating those cases.